lambda/tmc.ml

```diff
@@ -0,0 +1,854 @@
+    of the block containing the placeholder, and the offset of the
+    placeholder within the block.
+
+    In the common case, this offset is an arbitrary lambda expression, typically
+    a constant integer or a variable. We define ['a destination] as parametrized
+    over the offset type to represent formal destination parameters (where
+    the offset is an Ident.t), and maybe in the future statically-known
+    offsets (where the offset is an integer).
+*)
+
+let add_dst_params ({var; offset} : Ident.t destination) params =
+  (var, Pgenval) :: (offset, Pintval) :: params
+
+let add_dst_args ({var; offset} : offset destination) args =
+  Lvar var :: offset :: args
+
+let assign_to_dst {var; offset; loc} lam =
+  Lprim(Psetfield_computed(Pointer, Heap_initialization),
+        [Lvar var; offset; lam], loc)
+
+(** The type [Constr.t] represents a reified constructor with a single hole, which can
+    be either directly applied to a [lambda] term, or be used to create
+    a fresh [lambda destination] with a placeholder.
+ *)
+module Constr = struct
+  type t = {
+    tag : int;
+    flag: Asttypes.mutable_flag;
+    shape : block_shape;
+    before: lambda list;
+    after: lambda list;
+    loc : Debuginfo.Scoped_location.t;
+  }
+
+  let apply ?flag con t =
+    let flag = match flag with None -> con.flag | Some flag -> flag in
+    let block_args = List.append con.before @@ t :: con.after in
+    Lprim (Pmakeblock (con.tag, flag, con.shape), block_args, con.loc)
+
+  let tmc_placeholder = Lconst (Const_base (Const_int 0))
+  (* TODO consider using a more magical constant like 42, for debugging? *)
+
+  let with_placeholder con (body : lambda destination -> lambda -> lambda) : lambda =
+    let k_with_placeholder = apply ~flag:Mutable con tmc_placeholder in
+    let placeholder_pos = List.length con.before in
+    let placeholder_pos_lam = Lconst (Const_base (Const_int placeholder_pos)) in
+    let block_var = Ident.create_local "block" in
+    Llet (Strict, Pgenval, block_var, k_with_placeholder,
+          body { var = block_var; offset = placeholder_pos_lam ; loc = con.loc } (Lvar block_var))
+
+  (** We want to delay the application of the constructor to a later time.
+      This may move the constructor application below some effectful
+      expressions (if we move into a context of the form [foo;
+      bar_with_tmc_inside] for example), and we want to preserve the
+      evaluation order of the other arguments of the constructor.  So we bind
+      them before proceeding, unless they are obviously side-effect free. *)
+  let delay_impure : block_id:int -> t -> (t -> lambda) -> lambda =
+    let bind_list ~block_id ~arg_offset lambdas k =
+      let can_be_delayed =
+        (* Note that the delayed subterms will be used
+           exactly once in the linear-static subterm. So
+           we are happy to delay constants, which we would
+           not want to duplicate. *)
+        function
+        | Lvar _ | Lconst _ -> true
```

Comment:
A variable cannot be arbitrarilly delayed. if the variable is mutable, moving it will change the evaluation order.

Reply:
While this was true when this code was written, as a happy coincidence #10090 happens to fix that problem.

Reply:
Right, I forgot that one. So this should just wait for a rebase.